source/extensions/filters/http/jwt_authn/jwks_cache.cc

```diff
@@ -0,0 +1,129 @@
+#include "extensions/filters/http/jwt_authn/jwks_cache.h"
+
+#include <chrono>
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/config/datasource.h"
+
+#include "jwt_verify_lib/check_audience.h"
+
+using ::google::jwt_verify::Jwks;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+namespace {
+
+// Default cache expiration time in 5 minutes.
+constexpr int PubkeyCacheExpirationSec = 600;
+
+class JwksDataImpl : public JwksCache::JwksData, public Logger::Loggable<Logger::Id::filter> {
+public:
+  JwksDataImpl(const ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRule& jwt_rule)
+      : jwt_rule_(jwt_rule) {
+    std::vector<std::string> audiences;
+    for (const auto& aud : jwt_rule_.audiences()) {
+      audiences.push_back(aud);
+    }
+    audiences_ = std::make_unique<::google::jwt_verify::CheckAudience>(audiences);
+
+    const auto inline_jwks = Config::DataSource::read(jwt_rule_.local_jwks(), true);
+    if (!inline_jwks.empty()) {
+      const Status status = setKey(inline_jwks,
+                                   // inline jwks never expires.
+                                   std::chrono::steady_clock::time_point::max());
+      if (status != Status::Ok) {
+        ENVOY_LOG(warn, "Invalid inline jwks for issuer: {}, jwks: {}", jwt_rule_.issuer(),
+                  inline_jwks);
+      }
+    }
+  }
+
+  const ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRule& getJwtRule() const override {
+    return jwt_rule_;
+  }
+
+  bool areAudiencesAllowed(const std::vector<std::string>& jwt_audiences) const override {
+    return audiences_->areAudiencesAllowed(jwt_audiences);
+  }
+
+  const Jwks* getJwksObj() const override { return jwks_obj_.get(); }
+
+  bool isExpired() const override { return std::chrono::steady_clock::now() >= expiration_time_; }
+
+  Status setRemoteJwks(const std::string& jwks_str) override {
+    return setKey(jwks_str, getRemoteJwksExpirationTime());
+  }
+
+private:
+  // Get the expiration time for a remote Jwks
+  std::chrono::steady_clock::time_point getRemoteJwksExpirationTime() const {
+    auto expire = std::chrono::steady_clock::now();
+    if (jwt_rule_.has_remote_jwks() && jwt_rule_.remote_jwks().has_cache_duration()) {
+      const auto& duration = jwt_rule_.remote_jwks().cache_duration();
+      expire +=
+          std::chrono::seconds(duration.seconds()) + std::chrono::nanoseconds(duration.nanos());
```

Comment:
Probably cleaner to work with proto `DurationUtil`, e.g. `DurationUtil::durationToMilliseconds`.